### DISCUSSION TITLE
eglot-stay-out-of should take care completion-at-point-functions

### DISCUSSION THREAD

**Discussion by ssl19** [joaotavora/eglot#1093] [General]
Currently `(setq eglot-stay-out-of '(completion-at-point-functions))` has no effect on `completion-at-point-functions`. `eglot-completion-at-point` will always be added to `completion-at-point-functions` when `eglot` is enabled.

In my opinion, it is better to change this https://github.com/joaotavora/eglot/blob/e501275e06952889056268dabe08ccd0dbaf23e5/eglot.el#L1696 to

```elisp
(unless (eglot--stay-out-of-p 'completion-at-point-functions)
  (add-hook 'completion-at-point-functions #'eglot-completion-at-point nil t))
```
to keep the consistency of `eglot-stay-out-of`.
So that users won't have to `(remove-hook 'completion-at-point-functions #'eglot-completion-at-point t)` after they set `(setq eglot-stay-out-of '(completion-at-point-functions))`, when they want to use their own modified version of `eglot-completion-at-point`. For example:

```elisp
(defun my-eglot-capf ()
 "Eglot capf with non-exclusive property and cachebust."
 (cape-wrap-properties 'eglot-completion-at-point :exclusive (if corfu-mode 'no)))
```
A modified eglot corfu/cape backend with  non-exclusive property.

**Comment by joaotavora** [comment]
`eglot-stay-out-of` is for other preferences that don't use `add-hook`, not for this.

I recommend you use `eglot-managed-mode-hook`. You can do all kinds of tricks there.

To modify eglot-completion-at-point in such a superheroic way, you can also advise the function.  Warranty void, of course :-)

**Reply by ssl19** [reply]
> `eglot-stay-out-of` is for other preferences that don't use `add-hook`, not for this.

Thank you very much for the explanation！

> I recommend you use `eglot-managed-mode-hook`. You can do all kinds of tricks there.

I'll  try this 😆
